indexer/secure_index_builder.go

```diff
@@ -0,0 +1,94 @@
+package indexer
+
+import (
+	"bufio"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/binary"
+	"github.com/jxguan/go-datastructures/bitarray"
+	"golang.org/x/crypto/pbkdf2"
+	"hash"
+	"os"
+	"search/index"
+	"search/util"
+)
+
+// SecureIndexBuilder stores the essential information needed to build the
+// indexes for the documents.
+type SecureIndexBuilder struct {
+	numKeys      int                   // The number of keys.  This is also the number of PRFs.
+	keys         [][]byte              // The keys for the PRFs. Derived from the masterSecret and the salts.
+	hash         func() hash.Hash      // The hash function to be used for HMAC.
+	trapdoorFunc func(string) [][]byte // The trapdoor function for the words
+	size         uint64                // The size of each index, i.e. the number of buckets in the bloom filter.  Smaller size will lead to higher false positive rates.
+}
+
+// CreateSecureIndexBuilder instantiates a `secureIndexBuilder`.  Sets up the
+// hash function, and derives the keys from the master secret and salts by using
+// PBKDF2.  Finally, sets up the trapdoor function for the words.
+func CreateSecureIndexBuilder(h func() hash.Hash, masterSecret []byte, salts [][]byte, size uint64) *SecureIndexBuilder {
+	sib := new(SecureIndexBuilder)
+	sib.keys = make([][]byte, len(salts))
+	for index, salt := range salts {
+		sib.keys[index] = pbkdf2.Key(masterSecret, salt, 4096, 32, sha256.New)
+	}
+	sib.hash = h
+	sib.numKeys = len(salts)
```

Comment:
Ah I was thinking about needing to know that for serialization, but I was being stupid.

Fixed